File: 0006-zigzag-conversion.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        # 遍历然后按行存储
        # 关键点在于 反转 存储索引： flag = - flag

        if numRows <= 1:
            return s

        res = [""] * numRows

        i, flag = 0, -1

        for c in s:
            res[i] += c
            if i == 0 or i == numRows - 1: flag = - flag
            i += flag

        return "".join(res)

if __name__ == "__main__":
    print(Solution().convert("PAYPALISHIRING", 3))

File: 0006-zigzag-conversion.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
char * convert(char * s, int numRows){
    if (numRows < 2) {
        return s;
    }

    int length = 0;

    do {
        length++;
    } while (s[length-1] != '\0');

    char *res[numRows];

    for (int i = 0; i < numRows; i++) {
        res[i] = (char *)malloc(length * sizeof(char));
    }

    int i = 0, flag = -1;
    char * c = (char *)malloc(2 * sizeof(char));
    c[1] = '\0';

    while (*s != '\0') {
        c[0] = s[0];
        strcat(res[i], c);
        if (i == 0 || i == numRows - 1) {
            flag = -flag;
        }
        s++;
        i += flag;
    }
    free(c);
    char *result = (char *)malloc(length * sizeof(char));
    for(int i = 0; i < numRows; i++) {
        strcat(result, res[i]);
        free(res[i]);
    }
    return result;
}

int main(int argc, char *argv[]) {
    char *s = "PAYPALISHIRING";
    printf("%s\n", convert(s, 3));
    return 0;
}